File: Memento/Memento/Memento.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Memento
{
    public class Memento
    {
        private readonly string savedState;

        private Memento(string stateToSave)
        {
            savedState = stateToSave;
        }

        public class Originator
        {
            private string state;
            // The class could also contain additional data that is not part of the
            // state saved in the memento.

            public void Set(string state)
            {
                Console.WriteLine("Originator: Setting state to " + state);
                this.state = state;
            }

            public Memento SaveToMemento()
            {
                Console.WriteLine("Originator: Saving to Memento.");
                return new Memento(state);
            }

            public void RestoreFromMemento(Memento memento)
            {
                state = memento.savedState;
                Console.WriteLine("Originator: State after restoring from Memento: " + state);
            }
        }
    }    
}

File: Python/Memento.py
"""
Memento pattern example.
"""
class Originator:
    _state = ""

    def set(self, state: str) -> None:
        print(f"Originator: Setting state to {state}")
        self._state = state

    def save_to_memento(self) -> "Memento":
        return self.Memento(self._state)

    def restore_from_memento(self, m: "Memento") -> None:
        self._state = m.get_saved_state()
        print(f"Originator: State after restoring from Memento: {self._state}")

    class Memento:

        def __init__(self, state):
            self._state = state

        def get_saved_state(self):
            return self._state


saved_states = []
originator = Originator()
originator.set("State1")
originator.set("State2")
saved_states.append(originator.save_to_memento())

originator.set("State3")
saved_states.append(originator.save_to_memento())

originator.set("State4")
